man/winkeydaemon.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.40)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "WINKEYDAEMON 1"
.TH WINKEYDAEMON 1 "2020-01-26" "Winkeydaemon" "K1EL keyer driver"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
winkeydaemon \- Morse Code daemon for the Winkey hardware keyer module
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBwinkeydaemon\fR [\fB\-enmV\fR] [\fB\-d\fR \fIdevice\fR] [\fB\-b\fR \fIbaud\fR] [\fB\-p\fR \fIport\fR] [\fB\-q\fR \fIstring\fR] [\fB\-s\fR \fIspeed\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBwinkeydaemon\fR is a program written in Perl used to control the the K1EL
Winkey2 keyer and some others that offer the K1EL protocol emulation
such as the Hamgadgets Master Keyer 1.
It provides an interface which is compatible to \fBcwdaemon\fR, which means
it can be used in its place.
\&\fBwinkeydaemon\fR listens on a \s-1UDP\s0 socket and outputs commands to the K1EL
keyer on a serial port.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR \fIdevice\fR" 12
.IX Item "-d device"
Serial device to be used (default is \fI/dev/ttyUSB0\fR)
.IP "\fB\-b\fR \fIbaud\fR" 12
.IX Item "-b baud"
Serial baud rate (default is \fI1200\fR)
.IP "\fB\-p\fR \fIport\fR" 12
.IX Item "-p port"
\&\s-1UDP\s0 port (default is \fI6789\fR, the standard \fBcwdaemon\fR port)
.IP "\fB\-s\fR \fIwpm\fR" 12
.IX Item "-s wpm"
Speed (default is \fI24\fR \s-1WPM, 0\s0 is speed potentiometer)
.IP "\fB\-q\fR string" 12
.IX Item "-q string"
\&\s-1QRV\s0 string, transmitted on start-up
.IP "\fB\-e\fR" 12
.IX Item "-e"
Transmitted \s-1CW\s0 is echoed back on the client's \s-1UDP\s0 port (see: \fB\-p\fR)
.IP "\fB\-m\fR" 12
.IX Item "-m"
Turns off (\fImutes\fR) sidetone in host mode on version 2 and 3 keyers.
On \s-1USB\s0 Lite Wk2 keyers with no speaker, this may disable the \s-1LED.\s0
.IP "\fB\-n\fR" 12
.IX Item "-n"
Run in debug mode, stay in foreground
.IP "\fB\-V\fR" 12
.IX Item "-V"
Prints version, then exists.
.SH "CONFORMING TO cwdaemon"
.IX Header "CONFORMING TO cwdaemon"
A command consists of either a string of \s-1ASCII\s0 text to be sent as Morse
Code by the keyer, or a string consisting of a leading ESCape character
(decimal 27) followed immediately by a numeral or a single \s-1ASCII\s0 command
character and an additional numeric value, depending on the command.
.PP
\&\fBwinkeydaemon\fR accepts the following \fBcwdaemon\fR commands:
.IP "<\s-1ESC\s0>0" 16
.IX Item "<ESC>0"
Reset to default values
.IP "<\s-1ESC\s0>2\fIspeed\fR" 16
.IX Item "<ESC>2speed"
Set keying speed (5 ... 60 \s-1WPM\s0)
.IP "<\s-1ESC\s0>4" 16
.IX Item "<ESC>4"
Abort message
.IP "<\s-1ESC\s0>5" 16
.IX Item "<ESC>5"
Stop (exit) the daemon
.IP "<\s-1ESC\s0>7\fIweight\fR" 16
.IX Item "<ESC>7weight"
Set weighting (\-50 ... 50)
.IP "<\s-1ESC\s0>c\fIx\fR" 16
.IX Item "<ESC>cx"
Tune \fIx\fR seconds long (limit = 10 seconds)
.IP "<\s-1ESC\s0>d\fIdelay\fR" 16
.IX Item "<ESC>ddelay"
\&\s-1PTT\s0 on delay (0 ... 50ms)
.IP "\fIAny message\fR" 16
.IX Item "Any message"
Send Morse Code message
.IP "\fIqrz de pa0rct +test\-\fR" 16
.IX Item "qrz de pa0rct +test-"
Increase and decrease speed on the fly in 2 \s-1WPM\s0 steps (each \fB+\fR
and \fB\-\fR character in the string increases and decreases the
Morse speed by 2 \s-1WPM,\s0 for example, \fI+++test\-\-\-\fR will cause the
word \fItest\fR to be sent 6 \s-1WPM\s0 faster than the preceeding text (if
any) and the succeeding text (if any) will be sent at the set
speed).
.SH "NOTES"
.IX Header "NOTES"
The Perl module \fBDevice::SerialPort\fR is required. For Debian and derivatives
this can be found as the \fBlibdevice-serialport-perl\fR package.
For Slackware and derivatives this can be found on Slackbuilds dot org
(\fBhttp://slackbuilds.org\fR) as the perl-Device-SerialPort build script.
Other distributions will vary. It is always available from meta::cpan
(\fBhttps://metacpan.org/release/Device\-SerialPort\fR).
.PP
While an \fBinit\fR(8) script is included, running \fBwinkeydaemon\fR as a system
daemon is not required. A normal user account can run \fBwinkeydaemon\fR as
long as the account is a member of the group of the serial device
(often \fIdialout\fR but can vary).
.SH "BUGS"
.IX Header "BUGS"
A lack of testing on a wider range of keyers.
.PP
Bug reports and patches can be reported through the
<https://github.com/N0NB/winkeydaemon> GitHub Web site.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
Start \fBwinkeydaemon\fR using a custom serial device name:
.PP
winkeydaemon \-d /dev/mk1
.SH "AUTHORS"
.IX Header "AUTHORS"
\&\fBwinkeydaemon\fR was originally written by Rein Couperus, \s-1PA0R.\s0
.PP
Joop Stakenborg, \s-1PG4I,\s0 developed a package for the Debian GNU/Linux
distribution adding a manual page and an init script.
.PP
Wilbert Knol, \s-1PE7T,\s0 extended it to work with the Winkeyer \s-1USB\s0 models.
.PP
Nate Bargmann, N0NB, collated the various works and is hosting the project
at GitHub (<https://github.com/N0NB/winkeydaemon>).
.PP
This manual page was originally written by Joop Stakenborg, \s-1PG4I,\s0
for the Debian project (but may be used by others).
It has been expanded by Nate Bargmann, N0NB.


winkeydaemon.pl
#! /usr/bin/perl -w

# NAME
# winkeydaemon - morse daemon for the winkey hardware keyer module
#
# SYNOPSIS
# winkeydaemon [options] ...
#
# OPTIONS
# -n  Run in debug mode
# -p  UDP port (default is 6789)
# -s  Speed (default is 24 wpm, 0 is speed potentiometer)
# -d  Serial device (default is /dev/ttyUSB0)
# -q  QRV string, transmitted on start-up
# -V  Prints version, then exists.
#
# DESCRIPTION
#
# winkeydaemon.pl is a driver for the winkey2 keyer. It provides an interface
# which is compatible to the cwdaemon, which means it can be used instead of
# the cwdaemon.
# The winkeydaemon listens to a udp socket # and outputs commands to the k1el
# keyer on a serial port.
#

# The k1el_daemon implements the following cwdaemon-compatible commands:
# <ESC>"0"                 Reset to default values
# <ESC>"2"<"speed value">  Set keying speed (5 ... 60 wpm)
# <ESC>"4"                 Abort message
# <ESC>"5"                 Stop (Exit) the daemon
# <ESC>"7"<"weight value"> Set weighting (-50 ... 50)
# <ESC>"c"<"x">            Tune x seconds long (limit = 10 seconds)
# <ESC>"d"<"delay">        PTT on delay 0..50 (0 .. 50ms)
# Any message              Send morse code message  (max 1 packet!)
# qrz de pa0rct ++test--   In- and decrease speed on the fly in 4 wpm steps
#
# COPYING
#
# This program is published under the GPL license.
#   Copyright (C) 2007, 2008 Rein Couperus PA0R (rein at couperus.com)
#   Copyright (C) 2013 Wilbert Knol PE7T (knol00 at gmail.com)
#   Copyright (C) 2015 Nate Bargmann, N0NB <n0nb@arrl.net>
#
# *    winkeydaemon.pl is free software; you can redistribute it and/or modify
# *    it under the terms of the GNU General Public License as published by
# *    the Free Software Foundation; either version 2 of the License, or
# *    (at your option) any later version.
# *
# *    winkeydaemon.pl is distributed in the hope that it will be useful,
# *    but WITHOUT ANY WARRANTY; without even the implied warranty of
# *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# *    GNU General Public License for more details.
# *
# *    You should have received a copy of the GNU General Public License
# *    along with this program; if not, write to the Free Software Foundation,
# *    Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.

my $version = "version: 1.0.3   10 Dec 2015";
##
##    This is PA0R version 1.0 with modifications (below) by PE7T
##
##    To make  'winkeydaemon' work with  my 'K1EL winkeyer2  USB lite'
##    kitset, the following changes have been made:
##
## 1. Stop bits  has been changed to 2.
##
## 2. Initialisation  now starts with opening host  mode: '0x00 0x02'.
##    Reset '0x00 0x01' has been removed.
##
## 3. Iambic mode 'B' has been made the default: '0x0E 0x84'
##
## 4. The default serial port has been changed to '/dev/ttyUSB0'
##
## 5. Cleaned up spurious whitespaces and re-indented the code.
##
## 6. Fixed  some  (but   not  all)  'uninitialized  value'  run-time
##    warnings.
##
## 7. Changed  'buffered speed  change 0x1C' to  'set WPM  speed 0x02'
##    Reuesting a speed of '0'  selects the speed potentiometer as the
##    speed source.
##
## 8. 'Winkeyer USB  lite' will  not program  a requested  speed  that
##    falls  outside  the potentiometer  speed  range.  This range  is
##    defined  by  '$minSpeed'  and  '$maxSpeed'.  A  warning  is  now
##    printed on STDERR when the user requests an out-of-range speed.
##
## 9. Removed  automatic 'QRV'  on  start-up,  to prevent  accidental
##    QRM. Replaced with '-q <qrv-string>' option.
##
## 10 Configuring the serial  port using Device::SerialPort appears to
##    be buggy.  As a work-around, a  system call is made to 'stty' to
##    force 1200 baud, 8 databits, 2 stop bits, raw, no echo etc etc.

use Device::SerialPort qw( :PARAM :STAT 0.07 );
use IO::Socket;
use Getopt::Std;

my $myserver;
my $cnt = 0;
my $string = "";
my $debug = 0;
my $speed = 32;
my $minSpeed = 20;
my $maxSpeed = 40;
my $serial = "/dev/ttyUSB0";
my $server_port = 6789;

getopts("Vnp:s:d:q:");

if ($opt_V) {
    $opt_V = undef;
    print "\n$0 $version\n\n";
    exit(0);
}

if ($opt_n) {
    $debug = 1;
}

if ($opt_p) {
	$server_port = $opt_p;
}

if ($opt_s) {
	$speed = $opt_s;
}

if ($opt_d) {
	$serial = $opt_d;
}

if ($opt_q) {
	$qrvMsge = uc($opt_q);
}

if (-d "/tmp/.winkey") {
	# ok, no action required
} else {
	my $dir = "/tmp/.winkey";
	`mkdir "$dir"`;
	if ($debug) {print "Arranging mutex directory\n";}
}

########## Initialize the serial port


my @sttyCmd=qw(1200 raw clocal cread -crtscts cs8 cstopb -parenb -echo -echoe -echoctl -echok -echonl -echoprt -isig -iexten);
unshift @sttyCmd, ("stty", "-F", $serial);
print "Running @sttyCmd \n" if $debug;
system(@sttyCmd) == 0 or die;


my $port=Device::SerialPort->new($serial)
    || die "$0: could not construct port.\n";


$port->write_settings or die;
$port->baudrate(1200) || die "$0: could not set baud rate.\n";
$port->parity("none") || die "$0: could not set parity.\n";
$port->databits(8)    || die "$0: could not set data bits.\n";
$port->stopbits(2)    || die "$0: could not set stop bits.\n";

$port->dtr_active(1)  || die "$0: could not set DTR.\n";
$port->rts_active(0)  || die "$0: could not set RTS.\n";

my $timeout = 0;

$port->read_char_time(0);
$port->read_const_time(1);

########## Initialize the udp port

$server = IO::Socket::INET->new(LocalPort => $server_port, Proto => "udp")
    or die "Couldn't setup udp server on port $server_port : $@\n";

########### Initialize keyer

$openKeyer = sprintf("%c%c", 0x00, 0x02);	## open keyer interface
$count = $port->write($openKeyer);

select undef,undef,undef, 0.3;

## set mode: normal word space, no autospace, serial echo,
##           iambic-B,  no paddle echo , paddle watchdog.
$setmode = sprintf ("%c%c", 0x0E, 0x84);
$count = $port->write($setmode);

select undef,undef,undef, 0.3;

$setmode = sprintf ("%c%c", 0x03, 0x32); ## weighting: none = 0x32
$count = $port->write($setmode);

select undef,undef,undef, 0.3;

my $setpins = sprintf ("%c%c", 0x09, 0x0F); ## set pinout: LED=side-tone
$count = $port->write($setpins);

select undef,undef,undef, 0.2;

$senddelay = sprintf ("%c%c%c", 0x04, 0x01, 0x00); ## set PTT lead. tail [ms]
$count = $port->write($senddelay);

select undef,undef,undef, 0.2;

## set min  WPM and speed range:
my $r = $maxSpeed - $minSpeed;
$setRange = sprintf ("%c%c%c%c", 0x05, $minSpeed, $r, 0);
$count = $port->write($setRange);

$setspeed = sprintf ("%c%c", 0x02, $speed);  ## set the default speed
$count = $port->write($setspeed);

if ($opt_n) {
	$opt_n = 0;
	$debug = 1;
	Do_operations();	## do not fork, debug
} else {
	if (fork) {			## run as daemon
		exit;
	} else {
		for my $handle (*STDIN,*STDOUT, *STDERR) { # silent...
			open $handle, "+<", "/dev/null"
			or die "Cannot reopen $handle to /dev/null: $!";
		}
		Do_operations();
	}

}

exit;


########## Start operations ########
sub Do_operations {
####################################

    my $busy = 0;
    my $echo = "";

   $count = $port->write($qrvMsge) if defined $qrvMsge;

    while (1) {

	if ($cnt > 31) {
	    eval {
		local $SIG{ALRM} = sub { die "alarm time out" };
		alarm 1;

		$myserver = $server->recv($datagram, 32);

		alarm 0;
		1;  # return value from eval on normalcy
	    };
	    $cnt = 0;
	} else {
	    $cnt++;
	}

	if ($datagram) {
	    my @chars = split '', $datagram;
	    if (ord($chars[0]) == 27) {

		if ($chars[1] eq "2") {		## set speed
		    $speed = $chars[2];
		    $speed = $speed .  $chars[3] if defined $chars[3];
		    if ($speed != 0) {
			warn "Warning: can't set $speed WPM outside pot range: $minSpeed...$maxSpeed\n"
			    unless (($minSpeed <= $speed) && ($speed <= $maxSpeed));
		    }
		    $setspeed = sprintf ("%c%c", 0x02, $speed);
		    $count = $port->write($setspeed);
		    if ($debug) {print "setspeed=$speed\n";}
		} elsif ($chars[1] eq "5") {  ## exit daemon
		    last;
		} elsif ($chars[1] eq "7") {  ## set weight
		    if ($chars[2] eq "-") {
			$wgt = $chars[2] . $chars[3] ;
			if ( defined $chars[4] ) {
			    $wgt = $chars[2] . $chars[3] . $chars[4]
				unless ord($chars[4]) == 0;
			}
		    } else {
			    $wgt = $chars[2];
			    if ( defined $chars[3] ) {
				$wgt = $chars[2] . $chars[3]
				    unless ord($chars[3]) == 0;
			    }
		    }
		    $wgt += 50;
		    if ($wgt < 10) { $wgt = 10;}
		    if ($wgt > 90) { $wgt = 90;}

		    if ($debug) {print "weight=$wgt\n";}
		    my $setweight = sprintf ("%c%c", 0x03, $wgt);
		    $count = $port->write($setweight);
		} elsif ($chars[1] eq "d") {  ## set PTT lead in (00...50)
		    if (ord($chars[3]) == 0) {
			$delay = 0;
			$delaybyte = 0x02;
		    } else {
			$delay = $chars[2] . $chars[3];
			$delaybyte = int ($delay / 10);
				}

		    if ($debug) { print "PTT lead in = $delay\n";}

		    if ($delaybyte > 5) { $delaybyte = 5; }

		    $senddelay = sprintf ("%c%c%c", 0x04, $delaybyte, 0x00);
		    $count = $port->write($senddelay);

		} elsif ($chars[1] eq "c") {  ## TUNE
		    if ($tune_on) {
			$tune_on = 0;
			if ($debug) { print "Tune off\n";}
		    } else {
			$tune_on = 1;
			if ($debug) { print "Tune on\n";}
		    }
		    my $tuning = sprintf("%c%c", 0x0B, $tune_on);
		    $count = $port->write($tuning);

		} else {
		    $stopkeying = sprintf("%c", 0x0A);
		    $count = $port->write($stopkeying);
		}
		$datagram = "";

	    } else {
		foreach $c (@chars) {
		    $cr = ord($c);
		    if (($cr > 47 && $cr < 58) || ($cr > 64 && $cr < 91) || $cr == 32) { # only 0-9A-Z
			$string .= $c;
			$c = "";
		    } elsif ($cr == 39 || $cr == 41 || $cr == 47
			     || $cr == 58 || $cr == 60
			     || $cr == 61 || $cr == 62
			     || $cr == 64 || $cr == 63) {
			$string .= $c;
			$c = "";
		    } elsif ($cr == 38) {	## '&'
			my $chrs = sprintf ("%c%s%s", 0x1B, "A", "S");
			$string .= $chrs;
		    } elsif ($cr == 33) {	## '!'
			my $chrs = sprintf ("%c%s%s", 0x1B, "S", "N");
			$string .= $chrs;
		    } elsif ($cr == 40) {
			$string .= ")";
			$c = "";
		    } elsif ($cr == 42) {
			$string .= "<";
			$c = "";
		    } elsif ($cr == 43) {
			if ($speed < 90) {
			    $speed += 4;
			    $setspeed = sprintf ("%c%c", 0x1C, $speed);
			    $string .= $setspeed;
			}
		    } elsif ($cr == 45) {
			if ($speed > 8) {
			    $speed -= 4;
			    $setspeed = sprintf ("%c%c", 0x1C, $speed);
			    $string .= $setspeed;
			}
		    } elsif ($cr == 0) {
			last;
		    }
		    if ($debug > 1) {print $cr, "\n";}

		}

		if ($busy == 0) {
		    if (length ($string) > 30) {
			$outstring = substr($string, 0, 30);
			$string = substr($string, 30);
			$count = $port->write($outstring);
		    } elsif (length ($string) == 1) {
			$outstring = $string;
			$count = $port->write($outstring);
			$string = "";
		    } else {
			$outstring = $string;
			$string = "";
			$count = $port->write($outstring);
		    }
		} else {
		    $outstring = $string;
		    $string = "";
		    $count = $port->write($outstring);
		}

		if ($debug > 1) {print "WRITE:", $outstring, "\n";}

		$datagram = "";
	    }
	}

	($count,$saw)=$port->read(1); # will read 1 char
	if ($count) {
	    $stat = ord($saw);
	    if ($stat > 191) {
		if ($debug > 1) {print "\n", $stat, "\n";}
		if (($stat & 1) == 1) {
		    if ($debug){print "Buffer 2/3 full\n";}
		} elsif (($stat & 2) == 2) {
		    if ($debug){print "Brk-in\n";}
		} elsif (($stat & 4) == 4) {
		    if ($debug){print "Keyer busy\n";}
		    `touch /tmp/.winkey/keyer_busy`;
		    $busy = 1;
		    $echo = "";
		} elsif (($stat & 8) == 8) {
		    if ($debug){print "Tuning\n";}
		} elsif (($stat & 16) == 16) {
		    if ($debug){print "Waiting\n";}
		} else {
		    if ($debug){print "Idle\n";}
		    if (-e "/tmp/.winkey/keyer_busy") {
			`rm /tmp/.winkey/keyer_busy`;
		    }
		    $busy = 0;
		    $echo = "";
		}
	    } else {
		if ($debug) {print $saw, "\n";}

		if ($busy && $string) {

		    $echo .= $saw;
		    if (length($echo) > 9) {
			if (length($echo) <= length($string)) {
			    my $out = substr ($string, 0, length($echo));
			    $string = substr ($string, length ($echo));
			    $echo = "";
			    $count = $port->write($out);
			    $out = "";
			} else {
			    my $out = $string;
			    $echo = "";
			    $count = $port->write($out);
			    $string = "";
			}
		    } else {
			my $out = $string;
			$echo = "";
			$count = $port->write($out);
			$string = "";
		    }
		}
	    }
	}
    }

    $keyerclose = sprintf ("%c%c", 0x00, 3);
    $count = $port->write($keyerclose);
    undef $port;
    exit;
}

